### PR TITLE
Date input field constraints

### PR DESCRIPTION
**Pull request by KIRA009**
This PR adds the following constraints to the fields in the date input field
- `day` - 1 - 31
- `month` - 1 - 12
- `year` - 1 - 9999 (`datetime` package has limits `MINYEAR` and `MAXYEAR` for year)

This address #27 , and limits values of the inputs early on.

In addition, it also performs validation for the date as a whole, so incorrect dates like `31/04/2022` or `29/02/2022` will return a `ValidationError`, instead of failing at the `compress` method (when inserting these values to a `date` object) and causing a 500 error

`black` version has been updated, because of the error `ImportError: cannot import name '_unicodefun' from 'click'` encountered when running with `v22.1.0`. `v22.3.0` seems to have [fixed the issue](https://github.com/psf/black/issues/2964)

### CODE REVIEW

File: tbxforms/fields.py [hunk 1]
```diff
@@ -159,6 +204,20 @@ def clean(self, value):
                 field.widget.errors.extend(
                     m for m in e.messages if m not in field.widget.errors
                 )
+
+        # check if the given month and day are valid for the given year, this
+        # relies on the fields being declared in the order day, month, year
+        if len(clean_data) == 3:
+            day, month, year = map(int, clean_data)
+            dim = _days_in_month(year, month)
```
[the remark this answers is not in the document]
Comment (KIRA009) [reply tbxforms/fields.py:212]: Agreed, I just copy-pasted the logic used in the `datetime` package

File: tbxforms/validators.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from django.core.validators import (
+    MaxValueValidator,
+    MinValueValidator,
+)
+
+
+class StringMinValueValidator(MinValueValidator):
```
[the remark this answers is not in the document]
Comment (KIRA009) [reply tbxforms/validators.py:7]: Django doesn't actually, the value is passed on as string, and you get the `< not supported between instances of 'str' and 'int'` `TypeError` exception

File: tbxforms/fields.py [hunk 3]
```diff
@@ -1,12 +1,31 @@
...
+def _is_leap(year):
+    "year -> 1 if leap year, else 0."
+    return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)
+
+
+def _days_in_month(year, month):
+    "year, month -> number of days in that month in that year."
+    if month == 2 and _is_leap(year):
+        return 29
+    return _DAYS_IN_MONTH[month]
```
[the remark this answers is not in the document]
Comment (KIRA009) [reply tbxforms/fields.py:27]: I generally avoid using external functions which begin with `_`. The reasoning is that they are implementation details, which may change in the future and break things. Although in this case, it is a pretty simple one and not really subjected to change.

[the remark this answers is not in the document]
Comment (balazs-endresz) [reply tbxforms/fields.py:27]: Couldn't `_days_in_month(year, month)` be replaced with  `calendar.monthrange(year, month)[1]`?

> calendar.monthrange(year, month)
> Returns weekday of first day of the month and number of days in month, for the specified year and month.

File: tests/fields/test_date_input_field.py [hunk 4]
```diff
@@ -125,3 +125,44 @@ def test_invalid_year_field():
...
+    field = DateInputField()
+    with pytest.raises(ValidationError) as err:
+        field.clean(["11", "12", "999999999"])
+    error = err.value.args[0][0].args[0]
+    assert error in field.fields[2].widget.errors
+    assert error == field.fields[2].validators[2].message
+
+
+def test_invalid_date_on_leap_year():
+    """Verify an incorrect date on a leap year is not allowed."""
```
Comment (nimasmi) [inline tests/fields/test_date_input_field.py:161]: **Nitpick**: This is a test of non-leap year dates. Apparently they can be called [common years](https://en.wikipedia.org/wiki/Common_year), but I hadn't heard that before searching for it just now.

File: tbxforms/fields.py [hunk 5]
```diff
@@ -52,21 +71,47 @@ def __init__(self, **kwargs):
...
                 error_messages={"incomplete": _("Enter the month")},
                 validators=[
-                    RegexValidator(r"^[0-9]+$", _("Enter a valid month"))
+                    RegexValidator(r"^[0-9]+$", _("Enter a valid month")),
+                    StringMinValueValidator(
+                        1, _("Enter a value greater than 1")
+                    ),
+                    StringMaxValueValidator(
+                        12, _("Enter a value less than 12")
+                    ),
```
Comment (nimasmi) [inline tbxforms/fields.py:93]: **Suggestion**: Can we reword this to avoid the literal implication that 1 and 12 are not permitted, but still use the numbers 1 and 12 in the message?

E.g. "Enter a value not more than 12" or "Enter a value up to 12"?

- [Error Messages | GOV.UK Design System](https://design-system.service.gov.uk/components/error-message/#use-instructions-and-descriptions) suggests "Month must be 12 or less"
- [Date Input | GOV.UK Design System](https://design-system.service.gov.uk/components/date-input/#if-the-date-entered-cannot-be-correct) suggests "(field name) must be a real date", while highlighting just that sub field if it's just one field that's wrong, and highlighting the whole compound field if it's unclear or more than one is wrong

I reckon "Month must be 12 or less" etc. works well here.

NB this comment applies to Day and Year too.

File: tbxforms/fields.py [hunk 6]
```diff
@@ -159,6 +204,20 @@ def clean(self, value):
...
+
+        # check if the given month and day are valid for the given year, this
+        # relies on the fields being declared in the order day, month, year
+        if len(clean_data) == 3:
+            day, month, year = map(int, clean_data)
+            dim = _days_in_month(year, month)
+            if day > dim:
+                error = _(
+                    "Enter a day between 1 and %(dim)s for the month and year "
+                    "you entered"
```
Comment (nimasmi) [inline tbxforms/fields.py:216]: **Comment**: I think this message is clear, though maybe `from 1 to %(dim)s` would be more literally correct. "Between 1 and 30" also seems fine colloquially.

Note that the [Error Messages section of Date Input | GOV.UK Design System](https://design-system.service.gov.uk/components/date-input/#error-messages) has a lot to say on this matter and is worth a read.

File: tests/fields/test_date_input_field.py [hunk 7]
```diff
@@ -125,3 +125,44 @@ def test_invalid_year_field():
...
+    field = DateInputField()
+    with pytest.raises(ValidationError) as err:
+        field.clean(["29", "2", "2007"])
+    error = err.value.args[0][0]
+    assert error == (
+        "Enter a day between 1 and 28 for the month and year you entered"
+    )
```
Comment (nimasmi) [inline tests/fields/test_date_input_field.py:168]: **Suggestion**: Include a test of out-of-range day for a 30-day month, meaning for example 31 April.

Comment (nimasmi) [inline tests/fields/test_date_input_field.py:168]: **Suggestion**: Include a zero or negative case.

**Review by nimasmi — COMMENTED** [review]
In general, we should check the error messages follow the guidance at https://design-system.service.gov.uk/components/date-input/#error-messages. I have suggested some line-specific changes for this reason.

File: tbxforms/fields.py (hunk 5 above)
Comment (nimasmi) [reply tbxforms/fields.py]: This is resolved in db65d80.

File: tbxforms/fields.py (hunk 6 above)
Comment (nimasmi) [reply tbxforms/fields.py]: I think this is now a typo in db65d80: "between 1 to %(dim)s".

File: tbxforms/fields.py [hunk 8]
```diff
@@ -190,7 +190,7 @@ def clean(self, value):
             days_in_month = calendar.monthrange(year, month)[1]
             if day > days_in_month:
                 error = _(
-                    "Enter a day between 1 to %(dim)s for the month and year "
+                    "Enter a day from 1 to %(dim)s for the month and year "
                     "you entered"
                 ) % {"dim": days_in_month}
```
Comment (zerolab) [inline tbxforms/fields.py:195]:
```suggestion
                    "Enter a day from 1 to %(days_in_month)s for the month and year "
                    "you entered"
                ) % {"days_in_month": days_in_month}
```

surely this would make more sense when translating

Comment (KIRA009) [reply tbxforms/fields.py]: Fixed in [`449f085`](https://github.com/torchbox/tbxforms/pull/29/commits/449f0853079d1aaea40fb41aff78bfdeffff4033)

File: tbxforms/fields.py (hunk 6 above)
Comment (KIRA009) [reply tbxforms/fields.py]: Fixed in [`449f085`](https://github.com/torchbox/tbxforms/pull/29/commits/449f0853079d1aaea40fb41aff78bfdeffff4033)

File: tbxforms/fields.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KIRA009) [reply tbxforms/fields.py]: Added in [`0374597`](https://github.com/torchbox/tbxforms/pull/29/commits/0374597a0ec5c5c87a0f672b9e0afa4b8b02a634)

**Review by kbayliss — CHANGES_REQUESTED** [review]

File: tbxforms/fields.py [hunk 9]
```diff
@@ -159,6 +182,20 @@ def clean(self, value):
...
                 )
+
+        # check if the given month and day are valid for the given year, this
+        # relies on the fields being declared in the order day, month, year
+        if len(clean_data) == 3:
+            day, month, year = map(int, clean_data)
+            days_in_month = calendar.monthrange(year, month)[1]
+            if day > days_in_month:
+                error = _(
+                    "Enter a day from 1 to %(days_in_month)s for the month "
```
[the remark this answers is not in the document]
Comment (nimasmi) [reply tbxforms/fields.py:193]: Okay… This could be clearer.

File: tbxforms/fields.py [hunk 10]
```diff
@@ -52,21 +57,39 @@ def __init__(self, **kwargs):
                 label=_("Day"),
                 error_messages={"incomplete": _("Enter the day of the month")},
                 validators=[
-                    RegexValidator(r"^[0-9]+$", _("Enter a valid date"))
+                    RegexValidator(r"^[0-9]+$", _("Enter a valid date")),
```
Comment (kbayliss) [inline tbxforms/fields.py:60]: Looks like there's inconsistency with our error messages with some ending in full stops and others not (not as a result of this MR - looks like this has been an issue for a while).

Can we opt for ending with full stops, please?

File: tbxforms/fields.py [hunk 11]
```diff
@@ -52,21 +57,39 @@ def __init__(self, **kwargs):
...
+                    StringMinValueValidator(1, _("Day must be 1 or more")),
+                    StringMaxValueValidator(31, _("Day must be 31 or less")),
```
[the remark this answers is not in the document]
Comment (KIRA009) [reply tbxforms/fields.py:62]: Right, makes more sense to have a max length validator here, since down the line we are checking for a valid date anwyays.

File: tbxforms/fields.py (hunk 10 above)
Comment (KIRA009) [reply tbxforms/fields.py]: Added in 2fdcc2ae86fe279f9b523fdebafd5f9c44604c85

File: tbxforms/fields.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (KIRA009) [reply tbxforms/fields.py]: Changed in 2fdcc2ae86fe279f9b523fdebafd5f9c44604c85

File: tbxforms/fields.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (KIRA009) [reply tbxforms/fields.py]: Changed in 2fdcc2ae86fe279f9b523fdebafd5f9c44604c85

**Comment by lparsons396** [comment]
Hey, I was just wondering how things are going with this? If anyone has time to review @kbayliss' changes that would be amazing as there's an RNIB bug which could really do with this fix :)

**Comment by kbayliss** [comment]
This work has been merged as part of https://github.com/torchbox/tbxforms/pull/52 (it was easier to do within an integration branch due to so many related changes)
